libraries/typings-generator/src/TypingsGenerator.ts

```diff
@@ -80,89 +87,141 @@ export class TypingsGenerator {
 
     this._options.fileExtensions = this._normalizeFileExtensions(this._options.fileExtensions);
 
-    this._targetMap = new Map();
+    this._dependenciesOfFile = new Map();
+    this._consumersOfFile = new Map();
+    this._relativePaths = new Map();
 
-    this._dependencyMap = new Map();
+    this._fileGlob = `**/*+(${this._options.fileExtensions.join('|')})`;
   }
 
   public async generateTypingsAsync(): Promise<void> {
     await FileSystem.ensureEmptyFolderAsync(this._options.generatedTsFolder);
 
-    const filePaths: string[] = glob.sync(path.join('**', `*+(${this._options.fileExtensions.join('|')})`), {
+    const filePaths: string[] = glob.sync(this._fileGlob, {
       cwd: this._options.srcFolder,
       absolute: true,
       nosort: true,
       nodir: true
     });
 
-    for (let filePath of filePaths) {
-      filePath = path.resolve(this._options.srcFolder, filePath);
-      await this._parseFileAndGenerateTypingsAsync(filePath);
-    }
+    await this._reprocessFiles(filePaths);
   }
 
   public async runWatcherAsync(): Promise<void> {
     await FileSystem.ensureFolderAsync(this._options.generatedTsFolder);
 
-    const globBase: string = path.resolve(this._options.srcFolder, '**');
-
     await new Promise((resolve, reject): void => {
-      const watcher: chokidar.FSWatcher = chokidar.watch(
-        this._options.fileExtensions.map((fileExtension) => path.join(globBase, `*${fileExtension}`))
-      );
-      const boundGenerateTypingsFunction: (filePath: string) => Promise<void> =
-        this._parseFileAndGenerateTypingsAsync.bind(this);
-      watcher.on('add', boundGenerateTypingsFunction);
-      watcher.on('change', boundGenerateTypingsFunction);
-      watcher.on('unlink', async (filePath) => {
-        const generatedTsFilePath: string = this._getTypingsFilePath(filePath);
-        await FileSystem.deleteFileAsync(generatedTsFilePath);
+      const watcher: chokidar.FSWatcher = chokidar.watch(this._fileGlob, {
+        cwd: this._options.srcFolder
+      });
+
+      const queue: Set<string> = new Set();
+      let timeout: NodeJS.Timeout | undefined = undefined;
+      const flushQueue: () => void = () => {
+        const toProcess: string[] = Array.from(queue);
+        queue.clear();
+        timeout = undefined;
+        this._reprocessFiles(toProcess).catch(reject);
```

Comment:
Since this is kicking off an unawaited `Promise` on every change, and `this._reprocessFiles(...)` touches a handful of fields (`this._relativePaths`, `this._consumersOfFile`, and `this._dependenciesOfFile`), should we wait for the previous `this._reprocessFiles(...)` call to complete to avoid weird race-condition issues? It seems like you're using a 100ms debounce to dedupe file changes, which would help... I imagine my case would be rather rare, and would only occur if two batches large enough to take a long time to process _and_ processes the same file in both during processing that they collide.

Reply:
Good catch, do need to ensure the function is not reentrant.